docs/design/run.md

```diff
@@ -0,0 +1,268 @@
+
+
+**Summary:** This Document documents the design of a new `run` command for 
+Jujutsu which will run a user-provided command or script across multiple revisions. 
+This can be used for running build commands, linters or formatters. For more 
+details, read the [Use-Cases of jj run.](#Use-Cases-of-jj-run).
+
+## Preface
+
+The goal of this Design Document is to specify the correct behavior of `jj run`.
+The points we decide on here I (Philip Metzger) will try to implement. There 
+exists some prior work in other DVCS:
+* `git test`: part of [git-branchless]. Similar to this proposal for `jj run`. 
+* `hg run`: Google's internal Mercurial extension. Similar to this proposal for
+`jj run`.
+Details not available. 
+* `hg fix`: Google's open source Mercurial extension: [source code][fix-src]. A
+more specialized approach to rewriting file content without full context of the
+working directory. 
+* `git rebase -x`: runs commands opportunistically as part of rebase. 
+* `git bisect run`: run a command to determine which commit introduced a bug.
+
+## Context and Scope
+
+The initial need for some kind of command runner integrated in the VCS, surfaced
+in a [github discussion][pre-commit]. In a [discussion on discord][hooks] about
+the git-hook model, there was consensus about not repeating their mistakes.
+
+For `jj run` there is prior art in Mercurial, git branchless and Google's 
+internal Mercurial. Currently git-branchless `git test` and `hg fix` implement
+some kind of command runner. The Google internal `hg run` works in 
+conjunction with CitC (Clients in the Cloud) which allows it to lazily apply
+the current command to any affected file. Currently no Jujutsu backend
+(Git, Native) has a fancy virtual filesystem supporting it, so we 
+can't apply this optimization. We could do the same once we have an 
+implementation of the working copy based on a virtual file system. Until then, 
+we have to run the commands in regular local-disk working copies. 
+
+## Goals and Non-Goals
+
+### Goals
+
+* We should be able to apply the command to any revision, published or unpublished.
+* We should be able to parallelize running the actual command, while preserving a
+good console output.
+* The run command should be able to work in a working copy. 
+* There should exist some way to signal hard failure. 
+* The command should build enough infrastructure for `jj test`, `jj fix` and 
+`jj format`.
+* The main goal is to be good enough, as we can always expand the functionality 
+in the future.
+
+### Non-Goals
+
+* While we should build a base for `jj test`, `jj format` and `jj fix`, we 
+shouldn't mash their use-cases into `jj run`.
+* The command shouldn't be too smart, as too many assumptions about workflows 
+* makes the command confusing for users. 
+* Avoid the smart caching of outputs, as user input commands can be 
```

Comment:
(Optional): I'd delete this line about "Avoid smart caching", but up to you.

Reply:
I'm going to keep it, as the stated goal is to be good enough.